### ISSUE TITLE
sql: defer FK checks to end of statement

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#33475]
Discussed with @BramGruneir and team. The current semantics in CRDB 2.1 are incorrect -- the FK work must be deferred to no earlier than the end of the current statement.

In the first iteration of this we are reusing the same FK code; and instead delay the processing until the end of the statement's execution by means of callbacks.

Also the set of modified rows must be collected using a "disk row writer" that spills to disk to limit RAM usage.

**Comment by BramGruneir** [comment]
I agree that being able to delay the FK checks to the end of transaction would be really helpful, but for now, even moving them to the end of the statement, which should be the default, would be a big performance win over per row as we do them right now.  I expect cascading actions would be sped up here as well.

Once we're at that point, we can consider adding in the `deferred` keyword.

**Comment by vivekmenezes** [comment]
we discussed, I had forgotten that there is value to executing checks in parallel as statements are executing so as to 1. find violations earlier, 2. not deferring all check computations to the very end thereby slowing down a transaction.

**Comment by jordanlewis** [comment]
This will also get done as part of the recent opt FK work.

**Comment by RaduBerinde** [comment]
Opt-driven FK checks are enabled on master.
